polkadot/node/core/approval-voting/src/lib.rs

```diff
@@ -893,6 +942,93 @@ impl State {
 			},
 		}
 	}
+
+	fn mark_begining_of_gathering_assignments(
+		&mut self,
+		block_number: BlockNumber,
+		block_hash: Hash,
+		candidate: CandidateHash,
+	) {
+		let record = self
+			.per_block_assignments_gathering_times
+			.entry(block_number)
+			.or_default()
+			.entry((block_hash, candidate))
+			.or_default();
+
+		if record.stage_start.is_none() {
+			record.stage += 1;
+			gum::debug!(
+				target: LOG_TARGET,
+				stage = ?record.stage,
+				?block_hash,
+				?candidate,
+				"Started a new assignment gathering stage",
+			);
+			record.stage_start = Some(Instant::now());
+		}
+
+		// Make sure we always cleanup if we have too many records.
+		if self.per_block_assignments_gathering_times.len() >
+			MAX_BLOCKS_WITH_ASSIGNMENT_TIMESTAMPS as usize &&
+			block_number >= MAX_BLOCKS_WITH_ASSIGNMENT_TIMESTAMPS
+		{
+			self.cleanup_assignments_gathering_timestamp(
+				block_number - MAX_BLOCKS_WITH_ASSIGNMENT_TIMESTAMPS,
+			)
+		}
+	}
+
+	fn mark_gathered_enough_assignments(
+		&mut self,
+		block_number: BlockNumber,
+		block_hash: Hash,
+		candidate: CandidateHash,
+	) -> AssignmentGatheringRecord {
+		let record = self
+			.per_block_assignments_gathering_times
+			.get_mut(&block_number)
+			.and_then(|entry| entry.get_mut(&(block_hash, candidate)));
+		let stage = record.as_ref().map(|record| record.stage).unwrap_or_default();
+		AssignmentGatheringRecord {
+			stage,
+			stage_start: record.and_then(|record| record.stage_start.take()),
+		}
+	}
+
+	fn cleanup_assignments_gathering_timestamp(&mut self, keep_greater_than: BlockNumber) {
+		self.per_block_assignments_gathering_times
+			.retain(|block_number, _| *block_number > keep_greater_than)
```

Comment:
could probably be more efficient with a btree (split_off) or a sorted vecdeque (pop_front), but hopefully with this size (100), it should not be a big deal

Reply:
Given we prune this normall wouldn't be larger than a few items.

Reply:
Done, switched to LRUMap.